poh/src/poh_recorder.rs

```diff
@@ -196,6 +196,74 @@ pub enum PohLeaderStatus {
     Reached { poh_slot: Slot, parent_slot: Slot },
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub struct LeaderTickHeights {
+    pub first_tick_height: u64,
+    pub first_tick_reached_at: Option<Instant>,
+    pub target_tick_reached_at: Option<Instant>,
+    pub last_tick_height: u64,
+    pub grace_ticks: u64,
+}
+
+impl LeaderTickHeights {
+    fn new(first_tick_height: u64, last_tick_height: u64, grace_ticks: u64) -> Self {
+        Self {
+            first_tick_height,
+            first_tick_reached_at: None,
+            target_tick_reached_at: None,
+            last_tick_height,
+            grace_ticks,
+        }
+    }
+
+    // Returns the tick heights for when a validator next has its leader slots
+    fn compute_next_tick_heights(
+        next_leader_slots: Option<(Slot, Slot)>,
+        ticks_per_slot: u64,
+        current_tick_height: u64,
+    ) -> Option<Self> {
+        next_leader_slots.map(|(first_slot, last_slot)| {
+            let first_tick_height = first_slot * ticks_per_slot + 1;
+            let last_tick_height = (last_slot + 1) * ticks_per_slot;
+            let num_leader_slots = last_slot - first_slot + 1;
+            let grace_ticks = cmp::min(
+                ticks_per_slot * MAX_GRACE_SLOTS,
+                ticks_per_slot * num_leader_slots / GRACE_TICKS_FACTOR,
+            );
+
+            let mut tick_heights = Self::new(first_tick_height, last_tick_height, grace_ticks);
+            tick_heights.record_time_if_reached(current_tick_height);
+            tick_heights
+        })
+    }
+
+    fn record_time_if_reached(&mut self, current_tick_height: u64) {
+        let now = Instant::now();
+        let next_tick_height = current_tick_height.saturating_add(1);
+        if !self.first_tick_reached() && next_tick_height >= self.first_tick_height {
+            self.first_tick_reached_at = Some(now);
```

Comment:
nit: Can maybe just construct `Instant::now()` inside the if statements to avoid constructing the instant on every tick